ntoskrnl/se/token.c

```diff
@@ -135,34 +135,138 @@ SepDeleteTokenLock(
     ExFreePoolWithTag(Token->TokenLock, TAG_SE_TOKEN_LOCK);
 }
 
-static NTSTATUS
-SepCompareTokens(IN PTOKEN FirstToken,
-                 IN PTOKEN SecondToken,
-                 OUT PBOOLEAN Equal)
+/**
+ * @brief
+ * Compares tokens if they're equal given their properties.
+ *
+ * @param[in] FirstToken
+ * The first token.
+ *
+ * @param[in] SecondToken
+ * The second token.
+ *
+ * @param[out] Equal
+ * The retrieved value which determines if the tokens are
+ * equal or not.
+ *
+ * @return
+ * Returns STATUS_SUCCESS.
+ */
+static
+NTSTATUS
+SepCompareTokens(
+    _In_ PTOKEN FirstToken,
+    _In_ PTOKEN SecondToken,
+    _Out_ PBOOLEAN Equal)
 {
     BOOLEAN Restricted, IsEqual = FALSE;
+    ULONG FirstTokenIndex, SecondTokenIndex;
+    PAGED_CODE();
 
     ASSERT(FirstToken != SecondToken);
 
     /* Lock the tokens */
     SepAcquireTokenLockShared(FirstToken);
     SepAcquireTokenLockShared(SecondToken);
 
-    /* FIXME: Check if every SID that is present in either token is also present in the other one */
+    /* Check if every SID that is present in either token is also present in the other one */
+    for (FirstTokenIndex = 0; FirstTokenIndex < FirstToken->UserAndGroupCount; FirstTokenIndex++)
+    {
+        for (SecondTokenIndex = 0; SecondTokenIndex < SecondToken->UserAndGroupCount; SecondTokenIndex++)
+        {
+            if (!RtlEqualSid(FirstToken->UserAndGroups[FirstTokenIndex].Sid,
+                             SecondToken->UserAndGroups[SecondTokenIndex].Sid))
+            {
+                DPRINT1("SepCompareTokens(): SIDs are not equal!\n");
+                goto Quit;
+            }
+
+            /* Are the attributes from either SID from tokens equal? */
+            if (FirstToken->UserAndGroups[FirstTokenIndex].Attributes !=
+                SecondToken->UserAndGroups[SecondTokenIndex].Attributes)
+            {
+                DPRINT1("SepCompareTokens(): SID attributes are not equal!\n");
+                goto Quit;
+            }
+        }
+    }
```

Comment:
Is this expected that all the members of the UserAndGroups array are the same ?

I'd guess not. So please :

- Shortcut the loop if `FirstToken->UserAndGroupCount != SecondToken->UserAndGroupCount`
- Think again about why this is broken.

Reply:
> Is this expected that all the members of the UserAndGroups array are the same ?

To answer this question, a rostest that exactly builds two tokens that have some but not all such UserAndGroups attributes being the same, would be welcome.

Reply:
I made some research in regard of this and what Windows Server 2003 does is checks first if both the `UserAndGroupCount` counter member from either tokens are equivalent in order to proceed looping for SIDs and their attributes. Otherwise W2k3 simply bails out if the counters are not the same. Such case applies for privileges too.